music_shop/song/serializers.py

```diff
@@ -0,0 +1,88 @@
+from rest_framework import serializers
+
+from author.models import Author
+from author.serializers import AuthorSerializer
+from genre.models import Genre
+from genre.serializers import GenreSerializer
+from song.models import Song, SongData
+
+
+class SongDataSerializerGet(serializers.HyperlinkedModelSerializer):
+    class Meta:
+        model = SongData
+        fields = ("data",)
+
+
+class SongDataSerializerPost(serializers.ModelSerializer):
+    id = serializers.IntegerField()
+
+    class Meta:
+        model = SongData
+        fields = ("id",)
+
+
+class SongSerializerGet(serializers.ModelSerializer):
+    genre = GenreSerializer(many=True)
+    author = AuthorSerializer(many=True)
+    data = serializers.HyperlinkedRelatedField(
+        read_only=True,
+        view_name="song-data",
+    )
+
+    class Meta:
+        model = Song
+        fields = ("title", "author", "date", "genre", "data")
+
+
+class SongSerializerPost(serializers.ModelSerializer):
+    genre = GenreSerializer(many=True)
+    author = AuthorSerializer(many=True)
+    data = SongDataSerializerPost()
+
+    class Meta:
+        model = Song
+        fields = ("title", "author", "genre", "date", "data")
+
+    def create(self, validated_data):
+        genre = validated_data.pop("genre", [])
+        author = validated_data.pop("author", [])
+
+        data = validated_data.pop("data")
+        data_ = SongData.objects.get(pk=data.get("id"))
+        song = Song.objects.create(data=data_, **validated_data)
+
+        for current in genre:
+            genre_ = Genre.objects.get(pk=current.get("id"))
+            song.genre.add(genre_)
+
+        for current in author:
+            author_ = Author.objects.get(pk=current.get("id"))
+            song.author.add(author_)
+        return song
+
+    @staticmethod
+    def unpack_dict(dictionary):
+        for key, value in dictionary.items():
+            return value
```

Comment:
It is not the same, because `d.values()` returns an `odict_values` instance, which cannot be accessed by index. So to use `d.values()`, we have to convert it to list first to support random index access. The result will look like this:
```python
list(genre.values())[0]
```